### PR TITLE
Strict mocks support

### PR DESCRIPTION
**Pull request by mockitoguy**
Implemented support for strict mocks (#1097)

The PR is mergeable but I'd like to collect feedback first. What do you think?

### CODE REVIEW

File: src/main/java/org/mockito/quality/Strictness.java [hunk 1]
```diff
@@ -89,5 +94,49 @@
...
+     * If you are convinced you need {@code STRICT_MOCKS} mocking then read on.
+     * <p>
+     * This mode requires the developer to declare all stubbings for <strong>non-void</strong> methods
+     * invoked in the production code.
+     * Please remember that <strong>void methods</strong> are ignored and you are not required to stub them.
+     * <p>
+     * {@code STRICT_MOCKS} mocking is similar to traditional mocking with EasyMock,
+     * where undeclared invocations would fail early.
+     * {@code STRICT_MOCKS} mocking is useful for users migrating from EasyMock
+     * who are emotionally attached to stricter mocking.
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/quality/Strictness.java]: Let's not use words like `emotionally` here.
```suggestion
     * who prefer the behavior of stricter mocking.
```

File: src/main/java/org/mockito/quality/Strictness.java [hunk 2]
```diff
@@ -89,5 +94,49 @@
...
+     * clean tests and improved debuggability.
+     * {@code STRICT_MOCKS} mocking may increase brittleness of your tests
+     * and force you to declare additional stubbings that are irrelevant in given test method.
+     * This makes tests bigger and hence harder to read and comprehend.
+     * At times, {@code STRICT_MOCKS} mocking can be useful for troubleshooting tests or production code
+     * that is legacy or overcomplicated.
+     * <p>
+     * For discussion about {@code STRICT_MOCKS} mocking see ticket <a href="https://github.com/mockito/mockito/issues/1097">#1097</a>.
+     * <p>
+     * You can enable {@code STRICT_MOCKS} mocking via our JUnit4 support ({@link MockitoJUnit}) or {@link MockitoSession}
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/quality/Strictness.java]: The JUnit4 support is annotated with `@Deprecated` in this PR. If we want to deprecate it, we should not mention it in this javadoc.

**Comment by mockitoguy** [comment]
> I would prefer to tackle the stricts stubs issue before introducing more strict mocks support.

That's fair. What issue you have in mind?

> However, we have seen users running into issues with strict stubs and I fear strict mocks suffer the same fate.

By "issues" you mean bugs or known limitations of the API?

Thank you for reviewing!

File: src/main/java/org/mockito/quality/Strictness.java (hunk 1 above)
Comment (mockitoguy) [reply src/main/java/org/mockito/quality/Strictness.java]: +1

File: src/main/java/org/mockito/quality/Strictness.java (hunk 2 above)
Comment (mockitoguy) [reply src/main/java/org/mockito/quality/Strictness.java]: I deprecated only one specific JUnit4 runner subclass, not the entire JUnit4 support.

**Comment by mockitoguy** [comment]
Just run "./gradlew spotlessJavaApply" and it worked :-) Thanks!

**Comment by TimvdLippe** [comment]
> > I would prefer to tackle the stricts stubs issue before introducing more strict mocks support.
> 
> That's fair. What issue you have in mind?
> 
> > However, we have seen users running into issues with strict stubs and I fear strict mocks suffer the same fate.
> 
> By "issues" you mean bugs or known limitations of the API?
> 
> Thank you for reviewing!

I was mostly concerned with the "it reported this invocation was unused" in a test with a lot of setup code. I now realize that strict mocks are the inverse. You have to specify *everything*. So I think my initial concern was ungrounded.

I am okay with shipping this as `@Incubating` to collect user feedback.

**Comment by mockitoguy** [comment]
I'll close for now. Will reopen if we feel it is needed.

**Comment by andreisilviudragnea** [comment]
Why was this abandoned? The tests in our codebase highly benefited from the strict stubbing support and this was the next evolution we were waiting for, so the tests are even easier to understand and reason about.

**Comment by robertotru** [comment]
@mockitoguy hello, 
I hope you are doing well.

I'm wondering... why was this abandoned?

**Pull request closed without merging**